P5-jdk-version-support.md

```diff
@@ -0,0 +1,126 @@
+dependency. But when a transitive dependency was no longer supported (e.g.,
+because it has been replaced with a new major API version) the upgrade to a new
+Java version becomes a major undertaking. For library authors it was hard to
+know when it was "safe" to drop support for an old JDK version, and may find out
+the issues caused years after the decision. This created an ecosystem of sincere
+fear, uncertainty, and doubt where library owners did not want their
+dependencies to drop older JDK support, even though the benefit was unclear. It
+was rightly known that _some_ user was certainly still using an older JDK
+version, but it was unclear how many and whether they ever would need a new
+version of a library.
+
+That process has been changing since Java 8, especially with the new more rapid
+releases and planned LTS releases and reduced support timelines. Java releases
+have been offering features that have been encouraging upgrades and Java is
+hiding more of its internals to make upgrades easier. Running separate
+applications in a single, large application server is less common which allows
+upgrading pieces individually. The ecosystem has gotten more practiced at
+supporting new versions, less willing to be held back, and better at
+compartmentalizing legacy applications.
+
+gRPC is widely used and is lower in the dependency tree, so it has more
+conservative support. gRPC Java 1.0 supported Java 6, which was dropped
+following discussion in grpc/grpc-java#3961 because Java 6 was becoming
+unsupportable as the compiler was unmaintained and other libraries were dropping
+support. The discussion for dropping Java 7 is happening at grpc/grpc-java#4671
+and is necessary for reasons similar to Java 6. However, both processes were
+time-consuming and had no predictable timeline.
+
+gRPC has suffered from supporting old Java versions. There are hidden
+maintenance costs, but it is easy for someone to argue those should just be paid
+(either because of gRPC's scale or because they are not paying those costs).
+However, the public API surface has been limited which directly impacts users.
+The most obvious example of this is the lack of generated `interface`s for
+servers. Adding methods to an existing service must be backward-compatible, and
+the only way to do that before Java 8 is with an abstract class. But once we can
+use Java 8 language features we can use `default` methods in an `interface`.
+
+Android support is tracked separately and generally follows Google Play
+Services's supported API level, for security and as a reasonable guideline of
+what is worth supporting. Android support is not impacted by this gRFC as it has
+different options available to it (e.g., desugaring) and does not align cleanly
+to JDK versions.
+
+## Proposal
+
+gRPC Java may drop support for a JDK version when that version is no longer
+receiving [Premier Support from Oracle][Oracle Java support]. When doing so,
```

Comment:
> When doing so,

There will be two actions in the process:

- choose a minor version between the most recent versions of grpc for supporting the old JDK 
- make a new release which drops the old JDK

Which happens first? (I think the first one)  When do we notify the community when we starts the process? Can the gap time between the two be unlimited?

Reply:
> make a new release which drops the old JDK

Or we don't do this particularly?

Reply:
The smallest gap would be doing the two simultaneously from the external perspective. The release notes would say, "Java Y is now required. Non-Android users requiring Java Y-1, are encouraged to use v1.X".

When doing v1.X, I think we could say, "This is expected to be one of the last releases supporting Java Y and users requiring Java Y support are encouraged to use this release."

Right now defining the release after-the-fact is better because we haven't had a strong idea of when exactly we'd drop support. I had originally thought 1.36 would be a good candidate for Java 7 users, but it's now been 6 months and we still haven't dropped Java 7. With this policy planning is much easier so we could swap to the announce-ahead-of-time approach.

If you're asking if we could choose a year-old release, yes, we could try although it is easy to argue it isn't "recent." But I don't see much reason to do that. We want something recent because it'd already have the most features+bug fixes and less likely to require users to downgrade. But we may choose something slightly older because it has proved itself to have been a stable release. I think 6 months is the practical upper bound, with 3 months being more likely (which is just as a release would fall off the normal support window). Do you think that should be codified?

Reply:
> Do you think that should be codified?

No, it doesn't sound necessary.